### PR TITLE
Added class, linkText, partialLinkText and tagName strategies

### PR DESCRIPTION
**Pull request by email2vimalraj**
Added class, linkText, partialLinkText and tagName strategies.
Also updated unit tests to test the above strategies.

### CODE REVIEW

File: src/main/java/com/rationaleemotions/internal/locators/Strategy.java [hunk 1]
```diff
@@ -36,6 +36,62 @@ public By getStrategy(String locator) {
...
+     */
+    LINK_TEXT {
+        @Override
+        public boolean typeMatches(String locator) {
+            return Strategy.isNotNullAndEmpty(locator) && locator.startsWith("linkText");
+        }
+
+        @Override
+        public By getStrategy(String locator) {
+            return By.linkText(locator.substring(9));
```
Comment (krmahadevan) [inline src/main/java/com/rationaleemotions/internal/locators/Strategy.java]: This hard-coding of the substring length is kind of non intuitive, as in it doesn't really explain how we are arriving at this number. I guess we should instead being doing something like `By.linkText(locator.substring("linkText.")`.

We could even go the extra mile and define just two methods which each implementation of `typeMatches()` and `getStrategy()` would call into... Something like below.

```java
public boolean matches(String locator, String text) {
     return Strategy.isNotNullAndEmpty(locator) && locator.toLowerCase().startsWith(text.toLowerCase());
}
```

and then refactor the `typeMatches(locator)` to just call into this utility method.

The same thing with building a `getStrategy()` method.

You can perhaps just enhance the interface the enum is implementing to add a getter for the strategy text and then use that invocation to calculate the locator.

That way we aren't repeating anything any where and its all self explanatory.

Let me know if that explains what I am talking about.

In a nutshell, can we do the following :

1. Enhance `StrategyTraits` and add a `String locatorText()`
2. Create two utility methods such as 

```java
public boolean matches(String locator, String text) {
     return Strategy.isNotNullAndEmpty(locator) && locator.toLowerCase().startsWith(text.toLowerCase());
}
```
```java
private String extractLocator(String locator, String locatorText) {
   return locator.subString(locatorText.length() + 1);
}
```

and then have the `typeMatches("class")` call into `matches(locator, locatorText())` and do the similar thing for the other enum calls.

3. You now refactor `getStrategy(locator)` method to call into a utility method that would look like below

```java
public By getStrategy(String locator) {
   return By.linkText(locator, locatorText()); // here locatorText() is the method that comes from the interface `StategyTraits`
}
```

**Review by krmahadevan — CHANGES_REQUESTED** [review]
@email2vimalraj - Can you take of addressing the review comment that I shared, so that we can get this merged ?

**Comment by email2vimalraj** [comment]
Sure I'll make the changes accordingly and update.

**Comment by email2vimalraj** [comment]
@krmahadevan : First, apologies for the delay.

I've done the following things (**alert**: Breaking Change):

- Improved code as per the review feedback
- While doing this activity, found that there are few other ways to write xpath which was missing in the validation. For example, I could write my xpath as like this: `//(h1[@class='btn'])[1]` to identify the first matching element. I felt that instead of performing different checks, I made it more generic by following the same strategy pattern as others, like this: `xpath=//h1`.
- Updated the tests
- Updated the readme

Hope this holds good.

**Comment by krmahadevan** [comment]
@email2vimalraj 

> For example, I could write my xpath as like this: `//(h1[@class='btn'])[1]` to identify the first matching element. 

Why do you think this would not be parsed ? The current implementation already looks for `/` (or) `//` or `.`. The above path seems to satisfy this condition no ? I would like to understand more.

**Review by krmahadevan — APPROVED** [review]

**Comment by krmahadevan** [comment]
@email2vimalraj  - This is now released as version `1.0.3` http://repo1.maven.org/maven2/com/rationaleemotions/simple-se-pageobjects/1.0.3/

**Comment by email2vimalraj** [comment]
Thanks Krishnan.

**Pull request merged**